apm: Document tail-based sampling performance (#770)
elastic/docs-content: solutions/observability/apps/transaction-sampling.md

To demonstrate the performance overhead and requirements, here are some numbers from a standalone APM Server 9.0 deployed on AWS EC2, under full load receiving APM events containing only traces, assuming no backpressure from Elasticsearch, and 10% sample rate in tail sampling policy. They are for reference only, and may vary depending on factors like sampling rate, average event size, and average number of events per distributed trace.

| APM Server EC2 instance size | TBS enabled, Disk                                           | Event ingestion rate (throughput from APM agent to APM Server) in events/s | Memory usage (max Resident Set Size) in GB | Disk usage in GB |
|:-----------------------------|:------------------------------------------------------------|----------------------------------------------------------------------------|--------------------------------------------|------------------|

> | c6i.2xlarge or c6id.2xlarge  | TBS disabled                                                | 47220                                                                      | 0.98                                       | 0                |

Author: [to reviewer] it is 9.0 specific for now. I wonder if it makes sense to put in 8.18 numbers here (also to convince people to upgrade), as this is 9.0 docs.

| ..                           | TBS enabled, gp3 volume with the baseline IOPS of 3000 IOPS | 21310                                                                      | 1.41                                       | 13.1             |
| ..                           | TBS enabled, local NVMe SSD from c6id instance              | 21210                                                                      | 1.34                                       | 12.9             |

> | c6i.4xlarge or c6id.4xlarge  | TBS disabled                                                | 142200                                                                     | 1.12                                       | 0                |

Author: is it better to just say c6id.4xlarge instead of "c6i.4xlarge or c6id.4xlarge" to simplify the table? Now it is slightly hard to read. Not to mention the `..`

Author: took the liberty to change it

### Tail-based sampling performance and requirements [_tail_based_sampling_performance_and_requirements]

Tail-based sampling (TBS), by definition, requires storing events locally temporarily, such that they can be retrieved and forwarded once sampling decision is made.

In APM Server implementation, the events are stored temporarily on disk instead of in memory for better scalability. Therefore, it requires local disk storage proportional to APM event ingestion rate, and additional memory to facilitate disk reads and writes. Insufficient [storage limit](../../../solutions/observability/apps/transaction-sampling.md#sampling-tail-storage_limit) causes sampling to be bypassed.

It is recommended to use fast disks, such as NVMe SSDs, when enabling tail-based sampling. Disk throughput and I/O may become performance bottlenecks for tail-based sampling and APM event ingestion overall. Disk writes are proportional to the event ingest rate, while disk reads are proportional to both the event ingest rate and the sampling rate.

> To demonstrate the performance overhead and requirements, here are some reference numbers from a standalone APM Server deployed on AWS EC2 under full load, receiving APM events containing only traces. These numbers assume no backpressure from Elasticsearch and a 10% sample rate in the tail sampling policy. Please note that these figures are for reference only and may vary depending on factors such as sampling rate, average event size, and the average number of events per distributed trace.

Reviewer: ```suggestion
To demonstrate the performance overhead and requirements, here are some reference numbers from a standalone APM Server deployed on AWS EC2 under full load, receiving APM events containing only traces. These numbers assume no backpressure from Elasticsearch and a **10% sample rate in the tail sampling policy**. Please note that these figures are for reference only and may vary depending on factors such as sampling rate, average event size, and the average number of events per distributed trace.
```

Terminology:

* Event Ingestion Rate: The throughput from the APM agent to the APM Server using the Intake v2 protocol (the protocol used by Elastic APM agents), measured in events per second.
* Event Indexing Rate: The throughput from the APM Server to Elasticsearch, measured in events per second or documents per second.
* Memory Usage: The maximum Resident Set Size (RSS) of APM Server process observed throughout the benchmark.

| APM Server version | EC2 instance size | TBS and disk configuration                     | Event ingestion rate (events/s) | Event indexing rate (events/s) | Memory usage (GB) | Disk usage (GB) |
|--------------------|-------------------|------------------------------------------------|---------------------------------|--------------------------------|-------------------|-----------------|

> | 9.0                | c6id.2xlarge      | TBS disabled                                   | 47220                           | 4720 (100% sampling)           | 0.98              | 0               |

Reviewer: Is this right? It seems that the events/s ingestion rate is `47220` but the event indexing rate is 4720/s (10% of the events/s?

```suggestion
| 9.0                | c6id.2xlarge      | TBS disabled                                   | 47220                           | 47220 (100% sampling)           | 0.98              | 0               |
```

Author: Nice catch. I fat fingered it. I'll double check all the numbers before I merge.

Due to [OpenTelemetry tail-based sampling limitations](../../../solutions/observability/apps/limitations.md#apm-open-telemetry-tbs) when using [tailsamplingprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/tailsamplingprocessor), we recommend using APM Server tail-based sampling instead.

### Tail-based sampling performance and requirements [_tail_based_sampling_performance_and_requirements]

Tail-based sampling (TBS), by definition, requires storing events locally temporarily, such that they can be retrieved and forwarded once sampling decision is made.

In APM Server implementation, the events are stored temporarily on disk instead of in memory for better scalability. Therefore, it requires local disk storage proportional to APM event ingestion rate, and additional memory to facilitate disk reads and writes. Insufficient [storage limit](../../../solutions/observability/apps/transaction-sampling.md#sampling-tail-storage_limit) causes sampling to be bypassed.

> It is recommended to use fast disks, such as NVMe SSDs, when enabling tail-based sampling. Disk throughput and I/O may become performance bottlenecks for tail-based sampling and APM event ingestion overall. Disk writes are proportional to the event ingest rate, while disk reads are proportional to both the event ingest rate and the sampling rate.

Author: on second thought, spelling out SSD is probably good enough. Updated the sentence to mention ssd with high IOPS instead of NVMe which could be too specific

> To demonstrate the performance overhead and requirements, here are some reference numbers from a standalone APM Server deployed on AWS EC2 under full load, receiving APM events containing only traces. These numbers assume no backpressure from Elasticsearch and a **10% sample rate in the tail sampling policy**. Please note that these figures are for reference only and may vary depending on factors such as sampling rate, average event size, and the average number of events per distributed trace.

Reviewer 2: Maybe call out that the figures are only for reference in an `important` admonition? 

```suggestion
To demonstrate the performance overhead and requirements, here are some reference numbers from a standalone APM Server deployed on AWS EC2 under full load that is receiving APM events containing only traces. These numbers assume no backpressure from Elasticsearch and a **10% sample rate in the tail sampling policy**.

:::{important}
These figures are for reference only and may vary depending on factors such as sampling rate, average event size, and the average number of events per distributed trace.
:::
```

| 8.18               | c6id.2xlarge      | TBS disabled                                   | 50260                           | 50270 (100% sampling)          | 0.98              | 0               |
| 8.18               | c6id.2xlarge      | TBS enabled, EBS gp3 volume with 3000 IOPS     | 10960                           | 50                             | 5.24              | 24.3            |
| 8.18               | c6id.2xlarge      | TBS enabled, local NVMe SSD from c6id instance | 11450                           | 820                            | 7.19              | 30.6            |
| 8.18               | c6id.4xlarge      | TBS disabled                                   | 149200                          | 149200 (100% sampling)         | 1.14              | 0               |
| 8.18               | c6id.4xlarge      | TBS enabled, EBS gp3 volume with 3000 IOPS     | 11990                           | 530                            | 26.57             | 33.6            |
| 8.18               | c6id.4xlarge      | TBS enabled, local NVMe SSD from c6id instance | 43550                           | 2940                           | 28.76             | 109.6           |

> When interpreting these numbers, note that:
> 
> * The metrics are inter-related. For example, it is reasonable to see a higher memory usage and disk usage when event ingestion rate is higher.
> * Related to the previous point, event ingestion rate and event indexing rate competes for disk IO. It explains the outlier data point where 8.18 with a 32GB NVMe disk shows a higher ingest rate but a slower event indexing rate than in 9.0.

Reviewer 2: ```suggestion
When interpreting these numbers, note that:

* The metrics are inter-related. For example, it is reasonable to see higher memory usage and disk usage when the event ingestion rate is higher.
* The event ingestion rate and event indexing rate competes for disk IO. This is why there is an outlier data point where APM Server version 8.18 with a 32GB NVMe disk shows a higher ingest rate but a slower event indexing rate than in 9.0.
```
